File: src/main/java/de/cplaiz/activecraftcore/guicreator/GuiPageLayout.java
package de.cplaiz.activecraftcore.guicreator;

import de.cplaiz.activecraftcore.messagesv2.MessageSupplier;
import lombok.Getter;
import org.bukkit.entity.Player;
import org.bukkit.inventory.InventoryHolder;

@Getter
public abstract class GuiPageLayout {

    protected GuiPage guiPage;
    protected final String title;
    protected final int rows;
    protected final InventoryHolder holder;
    protected final String identifier;
    protected MessageSupplier messageSupplier;
    protected int currentPage;

    public GuiPageLayout(String identifier, int rows) {
        this(identifier, rows, null, GuiCreatorDefaults.guiTitle());
    }

    public GuiPageLayout(String identifier, int rows, InventoryHolder holder) {
        this(identifier, rows, holder, GuiCreatorDefaults.guiTitle());
    }

    public GuiPageLayout(String identifier, int rows, String title) {
        this(identifier, rows, null, title);
    }

    public GuiPageLayout(String identifier, int rows, String title, InventoryHolder inventoryHolder) {
        this(identifier, rows, inventoryHolder, title);
    }

    public GuiPageLayout(String identifier, int rows, InventoryHolder holder, String title) {
        this.title = title;
        this.rows = rows;
        this.holder = holder;
        this.identifier = identifier;
        this.guiPage = new GuiPage(this);
    }

    public abstract int getMaxPages();

    public abstract void refreshPage();

    public GuiPage navigateToPage(Player player, int pageNo) {
        currentPage = pageNo;
        GuiNavigator.pushReplacement(player, guiPage.build());
        return guiPage;
    }

    public GuiPage navigateToNextPage(Player player) {
        return navigateToPage(player, currentPage + 1);
    }

    public GuiPage navigateToPreviousPage(Player player) {
        return navigateToPage(player, currentPage - 1);
    }

}

File: src/main/java/org/activecraft/activecraftcore/guicreator/GuiPageLayout.kt
package org.activecraft.activecraftcore.guicreator

import org.activecraft.activecraftcore.guicreator.GuiCreatorDefaults.guiTitle
import org.activecraft.activecraftcore.guicreator.GuiNavigator.Companion.pushReplacement
import org.activecraft.activecraftcore.messages.MessageSupplier
import org.bukkit.entity.Player
import org.bukkit.inventory.InventoryHolder

abstract class GuiPageLayout constructor(
    val identifier: String,
    val rows: Int,
    val holder: InventoryHolder? = null,
    val title: String = guiTitle()
) {
    var guiPage: GuiPage = GuiPage(this)
    protected var messageSupplier: MessageSupplier? = null
    var currentPage = 0

    abstract val maxPages: Int
    abstract fun refreshPage()
    fun navigateToPage(player: Player?, pageNo: Int): GuiPage {
        currentPage = pageNo
        pushReplacement(player!!, guiPage.build())
        return guiPage
    }

    fun navigateToNextPage(player: Player?): GuiPage {
        return navigateToPage(player, currentPage + 1)
    }

    fun navigateToPreviousPage(player: Player?): GuiPage {
        return navigateToPage(player, currentPage - 1)
    }
}